Rung promotion limitation in ASHA/Hyperband to enable arbitrary unknown length runs

alexrobomind opened the pull request:
## Motivation
Currently, the ASHA and Hyperband pruners have no mechanism to limit the duration of a run by themselves. This is a problem when the length of a run is not capped by the user and the number of parallel processes is not high enough to afford wasting a few of them on runs that never terminate. This is particularly an issue in serial runs, as the initial run will just continue forever. Furthermore, when using hyperband in conjunction with the TPE sampler, the first runs which bootstrap TPE probably usually end up comparatively poor anyway, as the sampler operates fully randomly until bootstrap samples are collected.

This is however not inherently neccessary. This issue can be mitigated by just setting an expanding horizon for the run length, starting first with shorter runs and then expanding. It would be natural to include this expanding horizon into the pruner. As a side effect, this could make ASHA a more or less complete competitor for PBT, as one could now simply leave the scans running and let ASHA explore progressively deeper paths into the training space.

## Description of the changes
This pull request proposes a simple implementation strategy for the finite expanding horizon mentioned above. A new setting is introduced, which specifies a number of trials which must have terminated in a rung before any trial can be promoted into the next rung. At its default value (0), the ASHA and Hyperband pruners behave as usual. The proposed name for this setting is 'bootstrap_count', but that's just a suggestion.

The current implementation is incompatible with Hyperband's max_resource='auto', as the early stopping prohibits the pruner from determining max_resource as usual. This is acceptable, as this setting is primarily intended to be used with effectively infinite runs, where the run length is not predetermined. max_resource would therefore simply represent a point where Hyperband would transition into a usual ASHA pruner. One could potentially consider having other bracket scheduling techniques which would allow an infinite no. of brackets (e.g. exponential probability distribution over the brackets), but such considerations are beyond the scope of this PR.

HideakiImamura reviewed (changes requested):
Thanks for the PR! The introduced feature seems to be reasonable to me. This feature enables us to use Hyperband/ASHA with many-arms (rungs). I have several minor comments. Could you take a look?

[Just a note] The current Optuna's Hyperband cannot deal with infinite-arms(rungs), even if the feature introduced by this PR. I would like to know the idea to deal with it. @alexrobomind Do you have any idea?

alexrobomind commented on optuna/pruners/_successive_halving.py:
```diff
@@ -178,9 +189,16 @@ def prune(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial")
+            # 'competing' already includes the current trial
+            # Therefore, we need to use the '<=' operator here
+            if len(competing) <= self._bootstrap_count:
+                return True
```
This is supposed to be `True`. The logic behind this is, that, if there are too few competing runs (a.k.a. other runs and this run) that reached the current point, the computation should be aborted. I tested this and it works as intended.

HideakiImamura replied:
Thank you for the explanation. I misunderstood it.

alexrobomind commented on optuna/pruners/_hyperband.py:
```diff
@@ -152,6 +158,12 @@ def __init__(
+                "bootstrap_count > 0 and max_resource == 'auto' "
+                "are mutually incompatible, bootstrap_count is {}".format(self._bootstrap_count)
+            )
+
```
I was thinking that, since `bootstrap_count` passed to `SuccessiveHalvingPruner` it should be checked there to keep the checks in one case (so that if, for some reason, this restriction is lifted the hyperband pruner is transparently using the new strategy), but I can also add a check here. I will also look into the test cases.

HideakiImamura replied:
I agree with you. Could you add test cases?

alexrobomind commented on optuna/pruners/_hyperband.py:
```diff
@@ -152,6 +158,12 @@ def __init__(
             )
 
+        if self._bootstrap_count > 0 and self._max_resource == "auto":
+            raise ValueError(
+                "bootstrap_count > 0 and max_resource == 'auto' "
+                "are mutually incompatible, bootstrap_count is {}".format(self._bootstrap_count)
+            )
```
Will do.

alexrobomind commented on optuna/pruners/_successive_halving.py:
```diff
@@ -137,11 +141,18 @@ def __init__(
             )
 
+        if bootstrap_count < 0:
+            raise ValueError(
+                "The value of `bootstrap_count` is {}, "
+                "but must be `bootstrap_count >= 0`".format(bootstrap_count)
+            )
```
Same as above, will do that.

alexrobomind commented:
@HideakiImamura Thank you for reviewing this PR. I will implement the requested suggestions.

For doing (effectively) infinite-arm runs, I currently see two possible approaches:
- Option one is to deterministically sample the bracket id from a distribution over non-negative/positive integer numbers, e.g. an exponential distribution or a Poisson distribution.
- Option two is to expand the number of brackets over time, as the length of runs increases. This would have the advantage, that the "horizon" for the bracket calculation would always be finite, which would give more flexibility and a better controlled horizon expansion over the first strategy, which could potentially create very long runs already in the beginning. I personally favor this option.

The crux for making this work well is the time allocation strategy. The usual Hyperband approach, which is also (I think) what Optuna implements, is to distribute total computing time evenly along all brackets. Since the computing time required on average to make it through a bracket is approximately `length of longest rung * no. of rungs` (Hyperband is designed to spend a roughly even amount of time in each rung), this leads to the bracket with the fewest rungs hitting the "expansion point" first, which in turn would induce the creation of a new bracket. I also think that the goal of "fair distribution" of computing time does not really apply here anymore, as we are now dealing with an effectively infinite no. of brackets, which necessitates that we give some more time than others.

I think a suitable allocation strategy would be to determine the no. of brackets dynamically by setting `max_resource = max(resource count of all finished and pruned trials)` and then either give each bracket an even computing time with respect to that resource (current strategy, which will still over time move more and more resources to the older brackets, but they will always be playing catch-up) or sample from the brackets proportional to `reduction_factor ** rung_count_of_bracket`, which would on average get each bracket through this finite horizon at the same time.

Maybe Optuna should support always support both allocation strategies and set the default to the current one if the horizon is fixed and to the modified one if its dynamically expanding. In either case, however, the bracket allocation strategy would evolve over time, so the result of `_get_bracket_id` should be stored in the database and read from there.

alexrobomind commented:
@HideakiImamura I added the tests that you suggested. I also added a test that checks whether `boostrap_count` works as intended. I also found another conflict with the `min_resource=auto` case. That is now also checked for in the pruner and in the corresponding tests. Additionally, I squashed the commits together to have a cleaner history. Sorry for the noise this created in the PR history.

HideakiImamura commented:
@alexrobomind Sorry for the late review. I will restart the review now.

HideakiImamura reviewed (commented):
Thanks for adding tests. I have 2 minor comments on test cases. Could you take a look?

alexrobomind commented on tests/pruners_tests/test_hyperband.py:
```diff
@@ -211,3 +211,8 @@ def objective(trial: optuna.trial.Trial) -> float:
     study.optimize(objective, n_trials=10)
+
+
+def test_hyperband_boostrap_parameter() -> None:
```
Will change.

alexrobomind commented on tests/pruners_tests/test_successive_halving.py:
```diff
@@ -249,3 +249,26 @@ def test_successive_halving_pruner_min_early_stopping_rate_parameter() -> None:
+
+def test_successive_halving_pruner_bootstrap_parameter() -> None:
+
+    with pytest.raises(ValueError):
+        optuna.pruners.SuccessiveHalvingPruner(bootstrap_count=-1)
+
+    with pytest.raises(ValueError):
+        optuna.pruners.SuccessiveHalvingPruner(bootstrap_count=1, min_resource="auto")
```
This file has one test per parameter. I tried to follow the existing structure. If you prefer this change anyway, please confirm, then I will split the tests.

HideakiImamura replied:
Thanks for the confirmation. I understood your intention and agree with you. Let me resolve this conversation without any fix.

alexrobomind commented:
Changes implemented as requested. Thanks again for reviewing this PR.

HideakiImamura reviewed (approved):
Thanks for your swift actions! LGTM!

HideakiImamura commented:
@crcrpar You are originally the implementer of Hyperband. This PR is related to the implementation of Hyperband, so could you please review it?

crcrpar commented on optuna/pruners/_hyperband.py:
```diff
@@ -129,19 +129,25 @@ def objective(trial):
             See the details for :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        bootstrap_count:
+            Parameter specifying the number of trials required in a rung before any trial can be
+            promoted. Incompatible with max_resouce == 'auto'.
```
super nit

```suggestion
            promoted. Incompatible with ``max_resouce`` is ``"auto"``.
```

alexrobomind replied:
I admit that I don't quite understand this comment. I assume you are right, but I'm curious about the reasons.

- Are you refering to a style guide here (e.g. PEP 8)?
- Why use object identity (`is`) over object equality (`==`)?

crcrpar replied:
>     * Are you refering to a style guide here (e.g. PEP 8)?

No, if I understand what I do correctly. I said this from the perspective of how the documentation looks.  
I want every argument and variable to be double backquoted. With double backquotes, the value or expression like `max_resource / min_resource` rendered differently (orange characters) as in the capture below.
![image](https://user-images.githubusercontent.com/16191443/100956838-b97a5e80-355c-11eb-95d6-bf9ff70780b5.png)




>     * Why use object identity (`is`) over object equality (`==`)?

Object equality looks better :)

(I didn't mean to make this object identity by removing double quotes around "is" and I used "is" in order to make the sentence sounds natural when I read aloud.)

alexrobomind replied:
Ah, I see that I became a bit confused here. Thanks for explaining. I was just asking, because I am always looking for coding styles to reduce the style variance in my own programming.

crcrpar reviewed (commented):
Thanks for the pull request & apology for my delayed review.

This overall looks solid except for one nit-picking.
